### PR TITLE
INS-55: BaseProxy added

### PR DESCRIPTION
**Pull request by ilyapt**
**- What I did**
- Common proxy's methods moved to baseProxy
- Fixed interface of instanceDomain

**- How to verify it**
- Run tests

### CODE REVIEW

File: genesis/model/object/proxy.go [hunk 1]
```diff
@@ -20,3 +20,23 @@ package object
 type Proxy interface {
 	Child
 }
+
+// BaseProxy is a base implementation of Proxy.
+type BaseProxy struct {
+	Instance Child
+}
+
+// GetReference proxy call for instance method.
```
Comment (megge-dream) [inline genesis/model/object/proxy.go]: GetReference is a ...

File: genesis/model/object/proxy.go [hunk 2]
```diff
@@ -20,3 +20,23 @@ package object
...
+func (bp *BaseProxy) GetReference() *Reference {
+	return bp.Instance.GetReference()
+}
+
+// GetParent proxy call for instance method.
```
Comment (megge-dream) [inline genesis/model/object/proxy.go]: GetParent is a ...

File: genesis/model/object/proxy.go [hunk 3]
```diff
@@ -20,3 +20,23 @@ package object
...
+func (bp *BaseProxy) GetParent() Parent {
+	return bp.Instance.GetParent()
+}
+
+// GetClassID proxy call for instance method.
```
Comment (megge-dream) [inline genesis/model/object/proxy.go]: GetClassID is a ...

File: genesis/public/core/instance.go [hunk 4]
```diff
@@ -103,33 +103,20 @@ func newInstanceDomainProxy(parent object.Parent) (*instanceDomainProxy, error)
...
 	}
 	return &instanceDomainProxy{
-		instance: instance,
+		BaseProxy: object.BaseProxy{
+			Instance: instance,
+		},
 	}, nil
 }
 
 // CreateInstance proxy call for instance method.
```
Comment (megge-dream) [inline genesis/public/core/instance.go]: CreateInstance is a ...

File: genesis/public/core/instance.go [hunk 5]
```diff
@@ -103,33 +103,20 @@ func newInstanceDomainProxy(parent object.Parent) (*instanceDomainProxy, error)
...
 func (idp *instanceDomainProxy) CreateInstance(fc factory.Factory) (string, error) {
-	return idp.instance.CreateInstance(fc)
+	return idp.Instance.(*instanceDomain).CreateInstance(fc)
 }
 
 // GetInstance proxy call for instance method.
```
Comment (megge-dream) [inline genesis/public/core/instance.go]: GetInstance is a ...

File: genesis/public/core/reference.go [hunk 6]
```diff
@@ -106,44 +106,47 @@ func (rd *referenceDomain) ResolveReference(record string) (*object.Reference, e
...
 	return &referenceDomainProxy{
-		instance: newReferenceDomain(parent),
+		BaseProxy: object.BaseProxy{
+			Instance: newReferenceDomain(parent),
+		},
 	}
 }
 
-// RegisterReference is a proxy call for instance method.
+// RegisterReference proxy call for instance method.
```
Comment (megge-dream) [inline genesis/public/core/reference.go]: RegisterReference is a ...

File: genesis/public/core/reference.go [hunk 7]
```diff
@@ -106,44 +106,47 @@ func (rd *referenceDomain) ResolveReference(record string) (*object.Reference, e
...
 // ResolveReference is a proxy call for instance method.
 func (rdp *referenceDomainProxy) ResolveReference(record string) (*object.Reference, error) {
-	return rdp.instance.ResolveReference(record)
+	return rdp.Instance.(*referenceDomain).ResolveReference(record)
 }
 
 // GetReference is a proxy call for instance method.
-func (rdp *referenceDomainProxy) GetReference() *object.Reference {
-	return rdp.instance.GetReference()
+/*func (rdp *referenceDomainProxy) GetReference() *object.Reference {
```
Comment (megge-dream) [inline genesis/public/core/reference.go]: Why are this lines in the comments?

File: genesis/public/core/reference.go [hunk 8]
```diff
@@ -106,44 +106,47 @@ func (rd *referenceDomain) ResolveReference(record string) (*object.Reference, e
...
 // GetClassID is a proxy call for instance method.
 func (rdp *referenceDomainProxy) GetClassID() string {
 	return class.ReferenceDomainID
-}
+}*/
 
 // InitGlobalMap is a proxy call for instance method.
 func (rdp *referenceDomainProxy) InitGlobalMap(globalInstanceMap *map[string]object.Proxy) {
-	rdp.instance.InitGlobalMap(globalInstanceMap)
+	rdp.Instance.(*referenceDomain).InitGlobalMap(globalInstanceMap)
```
Comment (megge-dream) [inline genesis/public/core/reference.go]: May be it's better to convert rdp.Instance to interface ReferenceDomain?

File: genesis/public/core/reference_test.go [hunk 9]
```diff
@@ -116,7 +116,7 @@ func init() {
 	// Set map to ReferenceDomain.globalResolverMap
 	initRefDomain.InitGlobalMap(&globalResolverMap)
 	// Set map to ReferenceDomainProxy.ReferenceDomain.globalResolverMap
-	initRefDomainProxy.instance.InitGlobalMap(&globalResolverMap)
+	initRefDomainProxy.Instance.(*referenceDomain).InitGlobalMap(&globalResolverMap)
```
Comment (megge-dream) [inline genesis/public/core/reference_test.go]: Same, ReferenceDomain interface for convertion

File: genesis/public/core/instance.go [hunk 10]
```diff
@@ -93,7 +93,7 @@ func (instDom *instanceDomain) GetInstance(record string) (object.Proxy, error)
 }
 
 type instanceDomainProxy struct {
```
Comment (megge-dream) [inline genesis/public/core/instance.go]: instanceDomainProxy is not an implementation of InstanceDomain interface now, but it should be

File: genesis/public/core/reference.go [hunk 11]
```diff
@@ -106,44 +106,47 @@ func (rd *referenceDomain) ResolveReference(record string) (*object.Reference, e
 }
 
 type referenceDomainProxy struct {
```
Comment (megge-dream) [inline genesis/public/core/reference.go]: referenceDomainProxy is not an implementation of ReferenceDomain interface now, but it should be

File: genesis/public/core/reference.go (hunk 7 above)
Comment (ilyapt) [reply genesis/public/core/reference.go]: Ups(

File: genesis/public/core/class.go [hunk 12]
```diff
@@ -85,7 +86,8 @@ func (cd *classDomain) GetClass(recordID string) (factory.Factory, error) {
 }
 
 type classDomainProxy struct {
-	instance *classDomain
+	//instance *classDomain
```
Comment (megge-dream) [inline genesis/public/core/class.go]: Why is it comment?

File: genesis/public/core/class.go [hunk 13]
```diff
@@ -96,33 +98,21 @@ func newClassDomainProxy(parent object.Parent) (*classDomainProxy, error) {
 	}
 
 	return &classDomainProxy{
-		instance: instance,
+		//instance: instance,
```
Comment (megge-dream) [inline genesis/public/core/class.go]: Why is it comment?

File: genesis/public/core/reference.go [hunk 14]
```diff
@@ -173,8 +162,8 @@ func (rdf *referenceDomainFactory) GetReference() object.Reference {
 	return nil
 }
 
-// Create is a factory method for new ReferenceDomain instances.
-func (rdf *referenceDomainFactory) Create(parent object.Parent) (object.Proxy, error) {
+// Create factory method for new ReferenceDomain instances.
```
Comment (megge-dream) [inline genesis/public/core/reference.go]: Create is a ...

**Pull request merged**